lib/translator/unmap.cc

```diff
@@ -0,0 +1,92 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "unmap.h"
+
+namespace translator {
+
+// Section 5.6
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode UnmapToNvme(absl::Span<const uint8_t> scsi_cmd,
+                       absl::Span<const uint8_t> buffer_out,
+                       nvme::GenericQueueEntryCmd& nvme_cmd,
+                       Allocation& allocation, uint32_t nsid) {
+  scsi::UnmapCommand unmap_cmd;
+  if (!ReadValue(scsi_cmd, unmap_cmd)) {
+    DebugLog("Malformed unmap command");
+    return StatusCode::kFailure;
+  }
+  if (unmap_cmd.param_list_length < sizeof(scsi::UnmapParamList)) {
+    DebugLog("Insufficient unmap parameter list length");
+    return StatusCode::kFailure;
+  }
+  if (unmap_cmd.anchor == 1) {
+    DebugLog("Unsupported unmap anchor request");
+    return StatusCode::kNoTranslation;
+  }
+
+  // Copy unmap parameter list into memory
+  scsi::UnmapParamList param_list;
+  if (!ReadValue(buffer_out, param_list)) {
+    DebugLog("Malformed unmap parameter list");
+    return StatusCode::kFailure;
+  }
+  buffer_out = buffer_out.subspan(sizeof(scsi::UnmapParamList));
+
+  // Ensure that block descriptor data length & span length align
+  if (buffer_out.size() < param_list.block_desc_data_length) {
+    DebugLog("Block descriptor list length reported longer than buffer");
+    return StatusCode::kFailure;
+  }
+  uint32_t block_descriptor_count =
+      param_list.block_desc_data_length / sizeof(scsi::UnmapBlockDescriptor);
+  if (block_descriptor_count == 0 || block_descriptor_count > 256) {
+    DebugLog("Unsupported unmap block descriptor count %u",
+             block_descriptor_count);
+    return StatusCode::kNoTranslation;
+  }
+
+  // Copy block descriptors into structures
+  scsi::UnmapBlockDescriptor block_descriptors[block_descriptor_count];
+  for (uint32_t i = 0; i < block_descriptor_count; ++i) {
+    if (!ReadValue(buffer_out, block_descriptors[i])) {
+      DebugLog("Failed to read unmap block descriptor");
+      return StatusCode::kFailure;
+    }
+    buffer_out = buffer_out.subspan(sizeof(block_descriptors[i]));
+  }
+
+  // Copy block descriptor data into allocated nvme data buffer
+  if (allocation.SetPages(1, 0) == StatusCode::kFailure)
+    return StatusCode::kFailure;
+  nvme::DatasetManagmentRange* dmr_ptr =
+      reinterpret_cast<nvme::DatasetManagmentRange*>(allocation.data_addr);
+  for (uint32_t i = 0; i < block_descriptor_count; ++i) {
+    dmr_ptr[i].lba = block_descriptors[i].logical_block_addr;
+    dmr_ptr[i].lb_count = block_descriptors[i].logical_block_count;
+  }
+
+  // Create NVMe command
+  uint8_t bd_count_byte = static_cast<uint8_t>(block_descriptor_count - 1);
+  nvme::DatasetManagementCmd dataset_cmd = {
+      .opc = static_cast<uint8_t>(nvme::NvmOpcode::kDatasetManagement),
```

Comment:
why not make opc NvmOpcode in the struct def

Reply:
Other PRs in master have already used .opc = static_cast<uint8_t>(nvme::NvmOpcode::kDatasetManagement) so such a change would be dependent on a dedicated PR